various-receipes/3rd-party-services-isolation/test/isolation.test.js

```diff
@@ -0,0 +1,110 @@
+const request = require("supertest");
+const sinon = require("sinon");
+const nock = require("nock");
+const {
+  initializeWebServer,
+  stopWebServer
+} = require("../../../example-application/api-under-test");
+const mailer = require("../../../example-application/libraries/mailer");
+const OrderRepository = require("../../../example-application/data-access/order-repository");
+
+let expressApp;
+let sinonSandbox;
+
+beforeAll(async (done) => {
+  // ️️️✅ Best Practice: Isolate the service under test by intercepting requests to 3rd party services 
+  nock("http://localhost/user/").get(`/1`).reply(200, {
+    id: 1,
+    name: "John",
+  }).persist();
+
+  // ️️️✅ Best Practice: Place the backend under test within the same process
+  expressApp = await initializeWebServer();
+
+  // ️️️✅ Best Practice: use a sandbox for test doubles for proper clean-up between tests
+  sinonSandbox = sinon.createSandbox();
+
+  done();
+});
+
+afterAll(async (done) => {
+  // ️️️✅ Best Practice: Clean-up resources after each run
+  await stopWebServer();
+  done();
+});
+
+beforeEach(() => {
+  if (sinonSandbox) {
+    sinonSandbox.restore();
+  }
+});
+
+// ️️️✅ Best Practice: Structure tests 
+describe("/api", () => {
+  describe("POST /orders", () => {
+    test("When adding  a new valid order , Then should get back 200 response", async () => {
+      //Arrange
+      const orderToAdd = {
+        userId: 1,
+        productId: 2,
+        mode: "approved",
+      };
+
+      //Act
+      // ➿ Nock intercepts the request for users service as declared in the BeforeAll function
+      const orderAddResult = await request(expressApp).post("/order").send(orderToAdd);
+
+      //Assert
+      expect(orderAddResult.status).toBe(200);
+    });
+
+    test("When the user does not exist, return http 404", async () => {
+      //Arrange
+      // ️️️✅ Best Practice: Simulate 3rd party service responses to test different scenarios like 404, 422 or 500
+      nock("http://localhost/user/").get(`/7`).reply(404, {
+        message: "User does not exist",
+        code: "nonExisting",
+      });
+      const orderToAdd = {
+        userId: 7,
+        productId: 2,
+        mode: "draft",
+      };
+
+      //Act
+      const orderAddResult = await request(expressApp)
+        .post("/order")
+        .send(orderToAdd)
+
+      //Assert
+      expect(orderAddResult.status).toBe(404);
+    });
+
+    test("When order failed, send mail to admin", async () => {
+      //Arrange
+      process.env.SEND_MAILS = "true";
+      sinonSandbox.stub(OrderRepository.prototype, "addOrder").throws(new Error("Unknown error"));
+      // ️️️✅ Best Practice: Intercept requests for 3rd party services to eliminate undesired side effects like emails or SMS
+      // ️️️✅ Best Practice: Specify the body when you need to make sure you call the 3rd party service as expected 
```

Comment:
I'm not 100% sure about this - How specific one should be? Should any configurable text change make the tests fail or the important part is the fact that an email was sent? Maybe the middle is a good balance - Ensure that the subject and body have a non-empty string so we know that it is a valid email but don't care about the actual test?